File: backend/src/main/java/ua/kpi/its/lab/security/controller/TrainController.java
package ua.kpi.its.lab.security.controller;


import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import ua.kpi.its.lab.security.dto.CreateTrainRequest;
import ua.kpi.its.lab.security.dto.TrainDto;
import ua.kpi.its.lab.security.svc.TrainControllerService;

@RestController
@RequestMapping("/api/train")
public class TrainController {

    private final TrainControllerService trainService;

    @Autowired
    public TrainController(TrainControllerService trainService) {
        this.trainService = trainService;
    }

    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping
    public void createTrain(@RequestBody CreateTrainRequest createTrainRequest) {
        trainService.create(createTrainRequest);
    }

    @ResponseStatus(HttpStatus.OK)
    @GetMapping("/{id}")
    public TrainDto getTrain(@PathVariable("id") int id) {
        return trainService.buildTrainDtoFromTrainById(id);
    }

    @DeleteMapping("/{id}")
    public void deleteTrain(@PathVariable("id") int id) {
        trainService.deleteTrain(id);
    }

    @PatchMapping
    public TrainDto updateTrain(@RequestBody TrainDto trainDto) {
        return trainService.updateTrain(trainDto);
    }

    @GetMapping("/hello")
    public String getHello(){
        return "Hello World!";
    }

}

File: backend/src/main/kotlin/ua/kpi/its/lab/security/controller/TrainController.kt
package ua.kpi.its.lab.rest.controller

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*

import ua.kpi.its.lab.security.dto.TrainDto
import ua.kpi.its.lab.security.svc.TrainControllerService

/**
 * Контролер для управління поїздами.
 *
 * @property trainService Сервіс для обробки запитів по поїздах.
 */

@RestController
@RequestMapping("/api/train")
class TrainController @Autowired constructor(
    private val trainService: TrainControllerService
) {

    /**
     * Створює новий поїзд.
     *
     * @param createTrainRequest Запит на створення поїзда.
     */
    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping
    fun createTrain(@RequestBody createTrainRequest: TrainDto) {
        trainService.create(createTrainRequest)
    }

    /**
     * Отримує інформацію про поїзд за його ідентифікатором.
     *
     * @param id Ідентифікатор поїзда.
     * @return Інформація про поїзд у форматі [TrainDto].
     */
    @ResponseStatus(HttpStatus.OK)
    @GetMapping("/{id}")
    fun getTrain(@PathVariable("id") id: Int): TrainDto {
        return trainService.buildTrainDtoFromTrainById(id)
    }

    /**
     * Видаляє поїзд за його ідентифікатором.
     *
     * @param id Ідентифікатор поїзда, який потрібно видалити.
     */
    @DeleteMapping("/{id}")
    fun deleteTrain(@PathVariable("id") id: Int) {
        trainService.deleteTrain(id)
    }

    /**
     * Оновлює інформацію про поїзд.
     *
     * @param trainDto Інформація про поїзд для оновлення.
     * @return Оновлена інформація про поїзд у форматі [TrainDto].
     */
    @PatchMapping
    fun updateTrain(@RequestBody trainDto: TrainDto): TrainDto {
        return trainService.updateTrain(trainDto)
    }
}
